graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,204 @@
+                                           "Body is required");
+                }
+                final String query = (String) inputMap.getOrDefault("query", "");
+                final Map<String, Object> variables =
+                        (Map<String, Object>) inputMap.getOrDefault("variables", ImmutableMap.of());
+                final Map<String, Object> extensions =
+                        (Map<String, Object>) inputMap.getOrDefault("extensions", ImmutableMap.of());
+                return execute(ctx, GraphQLInput.builder(query, ctx)
+                                                .variables(variables)
+                                                .operationName((String) inputMap.get("operationName"))
+                                                .extensions(extensions)
+                                                .dataLoaderRegistry(dataLoaderRegistry)
+                                                .build());
+            }));
+        } else if (contentType != null &&
+                   contentType.type().equals(APPLICATION_GRAPHQL.type()) &&
+                   contentType.subtype().equals(APPLICATION_GRAPHQL.subtype())) {
+            return HttpResponse.from(request.aggregate().handleAsync((req, thrown) -> {
+                if (thrown != null) {
+                    logger.warn("{} Failed to aggregate a request:", ctx, thrown);
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(ctx, GraphQLInput.builder(query, ctx)
+                                                .dataLoaderRegistry(dataLoaderRegistry)
+                                                .build());
+            }));
+        } else {
+            return HttpResponse.of(HttpStatus.UNPROCESSABLE_ENTITY,
+                                   MediaType.PLAIN_TEXT,
+                                   "Could not process GraphQL request");
+        }
+    }
+
+    private HttpResponse execute(ServiceRequestContext ctx, GraphQLInput input) {
+        try {
+            final GraphQLOutput output = executor.serve(ctx, input);
+            return HttpResponse.of(MediaType.JSON_UTF_8,
+                                   deserialize(output.toSpecification()));
+        } catch (Exception e) {
+            final ExecutionResultImpl executionResult = new ExecutionResultImpl(
+                    GraphqlErrorException.newErrorException()
+                                         .message(e.getMessage())
+                                         .cause(e)
+                                         .build());
+            final GraphQLOutput output = GraphQLOutput.of(executionResult);
+            return HttpResponse.of(MediaType.JSON_UTF_8, deserialize(output.toSpecification()));
+        }
+    }
+
+    @Override
+    public Set<Route> routes() {
+        return ImmutableSet.of(route);
+    }
```

Comment:
I am not sure we need to implement `HttpServiceWithRoutes` for a `GraphQlService` which usually needs a single endpoint URL.

Reply:
I wanted to provide optional routing from the user's point of view.

```java
builder.service("/graphql", new GraphQLService() ...)
builder.service(GraphQLService.builder() ...)
```

In order to do that, I decided that we need to implement `HttpServiceWithRoutes`. The other way around is to provide separate methods, but I thought it was over-specified or dependent.

Is there any better way?

Reply:
> I wanted to provide optional routing from the user's point of view.

I thought users already can choose various routing options they want with the existing route API.

```java
Route route = Route.builder()
                    .path("/graphql")
                    .methods(...)
                    .build();
Server.builder()
      .service("/graphql", new GraphQlService())
      .service(route, new GraphQlService())
      .route().post(...)
              .build(new GraphQlService())
```
> The other way around is to provide separate methods

Could you elaborate a bit on separate methods? Does it mean HTTP methods or API design?

Reply:
If the path is something that is created from the `GraphQlService` when it's generated such as `GrpcService`, I think we don't need to use `HttpServiceWithRoutes`. 😄

Reply:
Since there is already a general usage(?) spec(?) for when using graphql over HTTP, I thought that it would be a bit unnecessary to set up a separate route when the user uses Armeria.

I'll still need a bit more refactoring, but for example, GraphQlService only uses specific HTTP methods (eg `get`, `post`). And post only allows specific media types(`application/json`, `application/graphql`, ... etc).

Reply:
> Could you elaborate a bit on separate methods? Does it mean HTTP methods or API design?

I mean API design.

Hmm... How about `GraphQlRoute` and `GraphQlService`? 🤔

Reply:
As discussed in Slack, we are going to remove `HttpServiceWithRoutes`.